core/src/main/java/org/opensearch/sql/analysis/Analyzer.java

```diff
@@ -448,6 +453,100 @@ public LogicalPlan visitEval(Eval node, AnalysisContext context) {
     return new LogicalEval(child, expressionsBuilder.build());
   }
 
+  /**
+   * Builds and returns a {@link org.opensearch.sql.planner.logical.LogicalFlatten} corresponding to
+   * the given flatten node.
+   */
+  @Override
+  public LogicalPlan visitFlatten(Flatten node, AnalysisContext context) {
+    LogicalPlan child = node.getChild().getFirst().accept(this, context);
+
+    ReferenceExpression fieldExpr =
+        (ReferenceExpression) expressionAnalyzer.analyze(node.getField(), context);
+    String fieldName = fieldExpr.getAttr();
+
+    // [A] Determine fields to add
+    // ---------------------------
+
+    // Iterate over all the fields defined in the type environment. Find all those that are
+    // descended from field that is being flattened. Determine the new path to add and remove the
+    // existing path. When determining the new path, we need to preserve the portion of the
+    // path corresponding to the flattened field's parent, if one exists, in order to support
+    // flattening nested structs - see example below.
+    //
+    // Input Data:
+    //
+    // [
+    //    struct: {
+    //      integer: 0,
+    //      nested_struct: { string: "value" }
+    //    }
+    // ]
+    //
+    // Example 1: 'flatten struct'
+    //
+    // [
+    //   integer: 0,
+    //   nested_struct: { string: "value" }
+    // ]
+    //
+    // Example 2: 'flatten struct.nested_struct'
+    //
+    // [
+    //    struct: {
+    //      integer: 0,
+    //      string: "value"
+    //    {
+    // ]
+
+    Map<String, ExprType> addFieldsMap = new HashMap<>();
+
+    TypeEnvironment env = context.peek();
+    Map<String, ExprType> fieldsMap = env.lookupAllTupleFields(Namespace.FIELD_NAME);
+
+    final String pathSeparator = ".";
+    final String fieldDescendantPath = fieldName + pathSeparator;
+    final Optional<String> fieldParentPath =
+        fieldName.contains(pathSeparator)
```

Comment:
should fieldName always be at the root?  For example, if we `flatten struct.nested_struct` and then there's a struct that contains `anotherobject.struct.nested_struct` this would match.

Reply:
oh, I didn't realize that you were doing another check for this later.